File: src/test/java/uk/gov/justice/hmpps/prison/repository/CaseLoadRepositoryTest.java
package uk.gov.justice.hmpps.prison.repository;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
import org.springframework.boot.test.autoconfigure.jdbc.JdbcTest;
import org.springframework.security.authentication.TestingAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;
import uk.gov.justice.hmpps.prison.api.model.CaseLoad;
import uk.gov.justice.hmpps.prison.web.config.PersistenceConfigs;

import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase.Replace.NONE;

@ActiveProfiles("test")
@Transactional(propagation = Propagation.NOT_SUPPORTED)
@JdbcTest
@AutoConfigureTestDatabase(replace = NONE)
@ContextConfiguration(classes = PersistenceConfigs.class)
public class CaseLoadRepositoryTest {
    private static final String TEST_USERNAME = "ITAG_USER";

    @Autowired
    private CaseLoadRepository repository;

    @BeforeEach
    public void init() {
        SecurityContextHolder.getContext().setAuthentication(new TestingAuthenticationToken(TEST_USERNAME, "password"));
    }

    @Test
    public void testGetCaseLoad() {
        final var caseLoad = repository.getCaseLoad("LEI");

        assertThat(caseLoad).get().extracting(CaseLoad::getDescription).isEqualTo("Leeds (HMP)");
    }

    @Test
    public void testGetAllCaseLoadsByUsername() {
        final var caseLoads = repository.getAllCaseLoadsByUsername(TEST_USERNAME);
        assertThat(caseLoads).extracting(CaseLoad::getCaseLoadId).containsOnly("LEI", "BXI", "MDI", "RNI", "SYI", "WAI", "NWEB");
    }

    @Test
    public void testGetCaseLoadsByUsername() {
        final var caseLoads = repository.getCaseLoadsByUsername(TEST_USERNAME);
        assertThat(caseLoads).extracting(CaseLoad::getCaseLoadId).containsOnly("LEI", "BXI", "MDI", "RNI", "SYI", "WAI");
    }

    @Test
    public void testGetCaseLoadsByUsername_CurrentActive() {
        final var caseLoads = repository.getCaseLoadsByUsername(TEST_USERNAME);
        final var activeCaseLoads = caseLoads.stream().filter(CaseLoad::isCurrentlyActive).collect(Collectors.toList());
        assertThat(activeCaseLoads).extracting(CaseLoad::getCaseLoadId).containsOnly("LEI");
    }

    @Test
    public void testGetCaseLoadsByUsername_CurrentInactive() {
        final var caseLoads = repository.getCaseLoadsByUsername(TEST_USERNAME);
        final var activeCaseLoads = caseLoads.stream().filter(cl -> !cl.isCurrentlyActive()).collect(Collectors.toList());
        assertThat(activeCaseLoads).extracting(CaseLoad::getCaseLoadId).containsOnly("BXI", "MDI", "RNI", "SYI", "WAI");
    }

    @Test
    public void testGetWorkingCaseLoadByUsername() {
        final var caseLoad = repository.getWorkingCaseLoadByUsername(TEST_USERNAME);

        assertThat(caseLoad).get().extracting(CaseLoad::getDescription).isEqualTo("Leeds (HMP)");
    }

}

File: src/test/java/uk/gov/justice/hmpps/prison/repository/CaseLoadRepositoryTest.kt
package uk.gov.justice.hmpps.prison.repository

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.jdbc.test.autoconfigure.AutoConfigureTestDatabase
import org.springframework.boot.jdbc.test.autoconfigure.JdbcTest
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.ContextConfiguration
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional
import uk.gov.justice.hmpps.prison.web.config.PersistenceConfigs
import uk.gov.justice.hmpps.test.kotlin.auth.WithMockAuthUser

@ActiveProfiles("test")
@Transactional(propagation = Propagation.NOT_SUPPORTED)
@JdbcTest
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
@ContextConfiguration(classes = [PersistenceConfigs::class])
@WithMockAuthUser("ITAG_USER")
class CaseLoadRepositoryTest(
  @Autowired private val repository: CaseLoadRepository,
) {
  @Test
  fun testGetCaseLoad() {
    val caseLoad = repository.getCaseLoad("LEI")

    assertThat(caseLoad).get().extracting { it.description }.isEqualTo("Leeds (HMP)")
  }

  @Test
  fun testGetAllCaseLoadsByUsername() {
    val caseLoads = repository.getAllCaseLoadsByUsername(TEST_USERNAME)
    assertThat(caseLoads)
      .extracting<String, RuntimeException> { it.caseLoadId }
      .containsOnly("LEI", "BXI", "MDI", "RNI", "SYI", "WAI", "NWEB")
  }

  @Test
  fun testGetCaseLoadsByUsername() {
    val caseLoads = repository.getCaseLoadsByUsername(TEST_USERNAME)
    assertThat(caseLoads)
      .extracting<String, RuntimeException> { it.caseLoadId }
      .containsOnly("LEI", "BXI", "MDI", "RNI", "SYI", "WAI")
  }

  @Test
  fun testGetCaseLoadsByUsername_CurrentActive() {
    val caseLoads = repository.getCaseLoadsByUsername(TEST_USERNAME)
    val activeCaseLoads = caseLoads.filter { it.isCurrentlyActive }
    assertThat(activeCaseLoads)
      .extracting<String, RuntimeException> { it.caseLoadId }
      .containsOnly("LEI")
  }

  @Test
  fun testGetCaseLoadsByUsername_CurrentInactive() {
    val caseLoads = repository.getCaseLoadsByUsername(TEST_USERNAME)
    val activeCaseLoads = caseLoads.filter { !it.isCurrentlyActive }
    assertThat(activeCaseLoads)
      .extracting<String, RuntimeException> { it.caseLoadId }
      .containsOnly("BXI", "MDI", "RNI", "SYI", "WAI")
  }

  @Test
  fun testGetWorkingCaseLoadByUsername() {
    val caseLoad = repository.getWorkingCaseLoadByUsername(TEST_USERNAME)

    assertThat(caseLoad).get().extracting { it.description }.isEqualTo("Leeds (HMP)")
  }

  companion object {
    private const val TEST_USERNAME = "ITAG_USER"
  }
}
